server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -400,18 +390,19 @@ private static void validateDynamicTemplate(Mapper.TypeParser.ParserContext pars
             return;
         }
 
-        final XContentFieldType[] types;
-        if (template.getXContentFieldType() != null) {
-            types = new XContentFieldType[]{template.getXContentFieldType()};
-        } else if (template.isRuntimeMapping()) {
-            types = Arrays.stream(XContentFieldType.values()).filter(XContentFieldType::supportsRuntimeField)
-                .toArray(XContentFieldType[]::new);
-        } else {
-            types = XContentFieldType.values();
+        if (parserContext.indexVersionCreated().before(Version.V_8_0_0)) {
+            if (template.pathMatch() == null && template.match() == null && template.getXContentFieldTypes().length == 0) {
+                throw new MapperParsingException("template [" + template.name() +
+                    " does not have [match] not [path_match] or [match_mapping_type] defined");
+            }
         }
 
+        final XContentFieldType[] types = Arrays.stream(template.getXContentFieldTypes())
+            .filter(t -> template.isRuntimeMapping() == false ||t.supportsRuntimeField())
+            .toArray(XContentFieldType[]::new);
+
         Exception lastError = null;
-        boolean dynamicTemplateInvalid = true;
+        boolean dynamicTemplateInvalid = types.length > 0;
```

Comment:
I am not following why this is needed, can you help me?

Reply:
I've simplified the existing validation code in https://github.com/elastic/elasticsearch/pull/69948/commits/00b92c15d6e66f8e6fe33953a03edf9107c31c3a.